File: Functional programming - OCaml/laboratory tasks/List5.ml
type point2D = float * float;;

let distance first second =
  match first with
  | (a, b) -> 
    match second with
    | (c, d) -> sqrt((a -. c)*.(a -. c) +. (b -. d)*.(b -. d))
;;

type point3D = float * float * float 

let distance3D first second =
  match first with
  | (a, b, c) -> 
    match second with
    | (d, e, f) -> sqrt((a -. d)*.(a -. d) +. (b -. e)*.(b -. e) +. (c -. f)*.(c -. f)) 
;;


type 'a listN = ('a)
let (xn : 'float listN) = [1.;1.;1.;1.;1.;1.];;

let distanceN newListN1 newListN2=
  let rec distanceNIn sum newListN1 newListN2 =
    match newListN1 with
    | [] -> sqrt(sum)
    | h1::t1 ->
      match newListN2 with
      | [] -> sqrt(sum)
      | h2::t2 -> distanceNIn (sum +. (h1 -. h2)*.(h1 -. h2)) (t1) (t2)
    in
    match newListN1 with
    | (list1) ->
      match newListN2 with
      | (list2) -> distanceNIn 0.0 list1 list2
;; 

(*Zadanie 2*)
(*Krotkowe rozwiązanie*)
type person = string * string * int * bool * int;;
type partnership = person * person;;

let (first : person) = "Jan", "Kowalski", 45, true, 42;;
let (second : person) = "Anna", "Kowalski", 46, false, 38;;

let (couple : partnership) = first, second;;

let giveMeYounger partners =
  match partners with 
  | (a, b) -> 
    match a with 
    | (aName, aSurname, aAge, aSex, aShoesNumber) ->
      match b with 
      | (bName, bSurname, bAge, bSex, bShoesNumber) ->
        if aAge > bAge then b
        else a
;;


(*Record rozwiązanie*)
type newPerson = {
    name : string;
    surname : string;
    age : int;
    sex : bool;
    shoesNumber : int;
  };;

type newPartnership = {
  first : newPerson;
  second : newPerson;
};;

let (jN : newPerson) = {name = "Jan"; surname = "Nowak"; age = 45; sex = true; shoesNumber = 42};;
let (aN : newPerson) = {name = "Anna"; surname = "Nowak"; age = 46; sex = false; shoesNumber = 38};;
let (couple : newPartnership) = {first = jN; second = aN};;

let newGiveMeYounger {first = a; second = b} =
  match a with 
   | {name = aName; surname = aSurname; age = aAge; sex = aSex; shoesNumber = aShoesNumber} ->
     match b with 
     | {name = bName; surname = bSurname; age = bAge; sex = bSex; shoesNumber = bShoesNumber} ->
       if aAge > bAge then b
       else a
;; 

(*Zadanie 3*)
type weekDay = Poniedzialek | Wtorek | Sroda | Czwartek | Piatek | Sobota | Niedziela;;

let weekDayToString day =
  match day with 
  | Poniedzialek -> "Poniedziałek"
  | Wtorek -> "Wtorek"
  | Sroda -> "Środa"
  | Czwartek -> "Czwartek"
  | Piatek -> "Piątek"
  | Sobota -> "Sobota"
  | Niedziela -> "Niedziela"
;;

let nextDay day =
  match day with 
  | Poniedzialek -> Wtorek
  | Wtorek -> Sroda
  | Sroda -> Czwartek
  | Czwartek -> Piatek
  | Piatek -> Sobota
  | Sobota -> Niedziela
  | Niedziela -> Poniedzialek
;;

(*Zadanie 4*)
type 'a maybe = Just of 'a | Nothing;;

let safeHead xs =
  match xs with
  | [] -> Nothing
  | h::t -> Just(h)
;;

(*Zadanie 5*)
type cuboid = float * float * float;;
type cone = float * float;;
type sphere = float;;
type cylinder = float * float;;

type solidFigue = Cuboid of cuboid | Cone of cone | Sphere of sphere | Cylinder of cylinder;;

let (p1 : cuboid) = 1.0, 1.0, 1.0;;
let (p2 : cone) = 1.0, 1.0;;
let (p3 : sphere) = 1.0;;
let (p4 : cylinder) = 1.0, 1.0;;
let (cuboid1 : solidFigue) = Cuboid p1;;
let (cone1 : solidFigue) = Cone p2;;
let (sphere1 : solidFigue) = Sphere p3;;
let (cylinder1 : solidFigue) = Cylinder p4;;

let volume figure =
  match figure with
  | Cuboid  (a, b, h)-> a*.b*.h
  | Cone (r, h)-> ((3.14 *. r *. r) *. h)/.3.0
  | Sphere (r) -> (4.0 *. 3.14 *. r *. r)/.3.0
  | Cylinder (r, h) -> (3.14 *. r *. r) *. h
;;

File: Functional programming - Scala/Laboratory tasks/Lista5/src/main/scala/List5.scala
import List5.{newPartnership, newPerson, partnership}

object List5 {

  //Zadanie 1
  type point2D = (Double, Double)

  def distance(first : point2D, second : point2D): Double = {
    first match {
      case (a, b) =>
        second match {
          case (c, d) => Math.sqrt(((a - c)*(a - c) + (b - d)*(b - d)))
        }
    }
  }

  type point3D = (Double, Double, Double)

  def distance3D(first: point3D, second: point3D): Double = {
    first match {
      case (a, b, c) =>
        second match {
          case (d, e, f) => Math.sqrt((a-d)*(a-d) + (b - e)*(b - e) + (c - f)*(c - f))
        }
    }
  }

  //Zadanie 2
  type person = (String, String, Int, Boolean, Int)
  type partnership = (person, person)

  def giveMeYounger(partners : partnership): person = {
    partners match {
      case (a, b) =>
        a match {
          case (aName, aSurname, aAge, aSex, aShoesNumber) =>
            b match {
              case (bName, bSurname, bAge, bSex, bShoesNumber) =>
                if (aAge > bAge) b;
                else a
            }
        }
    }
  }

  class newPerson(val name : String, val surName : String, val age : Int, val sex : Boolean, val shoesNumber : Int)
  class newPartnership(val first : newPerson, val second : newPerson)

  def newGiveMeYounger(partners : newPartnership): newPerson = {
    val first : newPerson = partners.first
    val second : newPerson = partners.second
    if first.age > second.age then second
    else first
  }

  //Zadanie 3
  enum weekDay():
    case Poniedzialek extends weekDay()
    case Wtorek extends weekDay()
    case Srode extends weekDay()
    case Czwartek extends weekDay()
    case Piatek extends weekDay()
    case Sobota extends weekDay()
    case Niedziela extends weekDay()

  def weekDayToString(day : weekDay): String = {
    day match {
      case weekDay.Poniedzialek => "Poniedziałek"
      case weekDay.Wtorek => "Wtorek"
      case weekDay.Srode => "Środa"
      case weekDay.Czwartek => "Czwartek"
      case weekDay.Piatek => "Piątek"
      case weekDay.Sobota => "Sobota"
      case weekDay.Niedziela => "Niedziela"
    }
  }

  def nextDay(day: weekDay): weekDay = {
    day match {
      case weekDay.Poniedzialek => weekDay.Wtorek
      case weekDay.Wtorek => weekDay.Srode
      case weekDay.Srode => weekDay.Czwartek
      case weekDay.Czwartek => weekDay.Piatek
      case weekDay.Piatek => weekDay.Sobota
      case weekDay.Sobota => weekDay.Niedziela
      case weekDay.Niedziela => weekDay.Poniedzialek
    }
  }

  //Zadanie 4
  enum maybe[A]():
    case Just(a : A) extends maybe[A]
    case Nothing() extends maybe[A]

  def safeHead[A](list : List[A]): maybe[A] = {
    list match {
      case Nil => maybe.Nothing()
      case h::t => maybe.Just(h)
    }
  }

  //Zadanie 5
  type cuboid = (Double, Double, Double)
  type cone = (Double, Double)
  type sphere = (Double)
  type cylinder = (Double, Double)
    enum solidFigure():
      case  Cuboid (c: cuboid) extends solidFigure
      case  Cone (c: cone) extends solidFigure
      case  Sphere (s : sphere) extends solidFigure
      case  Cylinder (c: cylinder) extends solidFigure

  def volume(figure: solidFigure): Double = {
    figure match {
      case solidFigure.Cuboid(p) =>
        val (a: Double, b: Double, h: Double) = p
        (a * b * h)
      case solidFigure.Cone(p) =>
        val (r: Double, h: Double) = p
        ((3.14 * r * r) * h) / 3.0
      case solidFigure.Sphere(p) =>
        val (r: Double) = p
        (4.0 * 3.14 * r * r) / 3.0
      case solidFigure.Cylinder(p) =>
        val (r: Double, h: Double) = p
        (3.14 * r * r * h)
    }
  }

  @main def main(): Unit = {
    val x1 : point2D = (1.0, 1.0)
    val x2 : point2D = (0.0, 0.0)
    println(distance(x1, x2))

    val x3 : point3D = (1.0,1.0,1.0)
    val x4 : point3D = (0.0,0.0,0.0)
    println(distance3D(x3, x4))

    //Zadanie 2
    val first : person = ("Jan", "Kowalski", 45, true, 42)
    val second : person = ("Anna", "Kowalski", 46, false, 38)
    val couple : partnership = (first, second)

    println(giveMeYounger(couple))

    val newFirst: newPerson = newPerson("Jan", "Kowalski", 45, true, 42)
    val newSecond: newPerson = newPerson("Anna", "Kowalski", 46, false, 38)
    val newCouple: newPartnership = newPartnership(newFirst, newSecond)

    println(newFirst == newGiveMeYounger(newCouple))


    //Zadanie 3
    val day1 = weekDay.Poniedzialek

    println(weekDayToString(day1))
    println(nextDay(day1))
    //Zadanie 4
    println(safeHead[Int](List(1,1,1,1,1)))
    println(safeHead[Int](List()))
    //Zadanie 5
    val par = (1.0d,1.0d,1.0d,1.0d)

    val c : cuboid = (1.0d, 1.0d, 1.0d)
    val figure1 : solidFigure = solidFigure.Cuboid(c)

    val co : cone = (1.0d, 1.0d)
    val figure2 : solidFigure = solidFigure.Cone(co)

    val s : sphere = (1.0d)
    val figure3 : solidFigure = solidFigure.Sphere(s)

    val cy : cylinder = (1.0d, 1.0d)
    val figure4 : solidFigure = solidFigure.Cylinder(cy)

    println(volume(figure1))
    println(volume(figure2))
    println(volume(figure3))
    println(volume(figure4))
  }
}
